trieste/bayesian_optimizer.py

```diff
@@ -206,8 +209,40 @@ def optimize(
         self,
         num_steps: int,
         datasets: Mapping[str, Dataset],
-        model_specs: Mapping[str, ModelSpec],
-        acquisition_rule: AcquisitionRule[State[S | None, TensorType], SP],
+        # there's no way to statically check config-based models
+        model_specs: Mapping[str, ModelConfigType],
+        acquisition_rule: AcquisitionRule[TensorType, SP, M_contra],
+        *,
+        track_state: bool = True,
+        fit_initial_model: bool = True,
+        # this should really be OptimizationResult[None], but tf.Tensor is untyped so the type
+        # checker can't differentiate between TensorType and State[S | None, TensorType], and
+        # the return types clash. object is close enough to None that object will do.
+    ) -> OptimizationResult[object]:
+        ...
+
+    @overload
+    def optimize(
+        self,
+        num_steps: int,
+        datasets: Mapping[str, Dataset],
+        model_specs: Mapping[str, M_contra],
+        acquisition_rule: AcquisitionRule[State[S | None, TensorType], SP, M_contra],
+        acquisition_state: S | None = None,
+        *,
+        track_state: bool = True,
+        fit_initial_model: bool = True,
+    ) -> OptimizationResult[S]:
+        ...
+
+    @overload
+    def optimize(
+        self,
+        num_steps: int,
+        datasets: Mapping[str, Dataset],
+        # there's no way to statically check config-based models
```

Comment:
`TypedDict`?

also, hi uri. This was a PR I was thinking about adding back when. Glad to see it's happening

Reply:
We may be able to do a bit better, but we'll always be limited since the type (and hence acquisition function compatibility) of a config-based model might only be inferable at runtime.

PS Feel free to leave lots more comments :-)

Reply:
it would make sense to use typed dictionaries since they always have to have same arguments - would that make it possible to do type checks?

(we should probably use typed dicts anyway...)

thanks @joelberkeley